log-parser/cpp/log_parser_hs.cpp

```diff
@@ -0,0 +1,619 @@
+		printf("[found] %s\n", sr.line.c_str());
+	}
+	*/
+	printf("[info] %zd results.\n", results.size());
+}
+
+int main(int argc, char* argv[]) {
+#if defined(LP_STL_UNORDERED)
+	printf("[config] stl unordered.\n");
+#elif defined(LP_STL_ORDERED)
+	printf("[config] stl ordered.\n");
+#elif defined(LP_ABSL_NODE)
+	printf("[config] absl node.\n");
+#elif defined(LP_ABSL_FLAT)
+	printf("[config] absl flat.\n");
+#else
+	printf("[config] badness.\n");
+	return 1;
+#endif
+
+#if defined(LP_REGEX_STD)
+	printf("[config] regex std.\n");
+#elif defined(LP_REGEX_BOOST)
+	printf("[config] regex boost.\n");
+#elif defined(LP_REGEX_HYPERSCAN)
+	printf("[config] regex hyperscan.\n");
+#else
+	printf("[config] badness.\n");
+	return 1;
+#endif
+
+#if defined(LP_O2)
+	printf("[config] -O2\n");
+#elif defined(LP_O3)
+	printf("[config] -O3\n");
+#else
+	printf("[config] badness.\n");
+	return 1;
+#endif
+
+	if (argc <= 4) {
+		printf("[usage] ./log_parser <num threads> <searches> <files> <indexed|regex|spooky>\n");
+		return 1;
+	}
+	size_t num_threads = std::strtoul(argv[1], nullptr, 10);
+	std::string type(argv[4]);
+	std::vector<std::string> searches;
+	{
+		std::ifstream f(argv[2]);
+		std::string line;
+		while (std::getline(f, line)) {
+			searches.push_back(line);
+		}
+	}
+	std::vector<std::string> files;
+	{
+		std::ifstream f(argv[3]);
+		std::string line;
+		while (std::getline(f, line)) {
+			files.push_back(line);
+		}
+	}
+
+	struct timespec t0, t1;
+	LogParser lp(files);
+	clock_gettime(CLOCK_MONOTONIC, &t0);
+	lp.index(num_threads);
+	clock_gettime(CLOCK_MONOTONIC, &t1);
+	printf("[info] indexing: %ld\n", time_diff(t0, t1));
+	if (type == "indexed") {
+		for (size_t i = 0; i < searches.size(); i++) {
+			std::string& s = searches[i];
+			printf("[info] indexed search %zd: %s\n", i, s.c_str());
+			std::vector<UniquePtr<LogParser::SearchResult>> results;
+			clock_gettime(CLOCK_MONOTONIC, &t0);
+			lp.search(num_threads, s, results);
+			clock_gettime(CLOCK_MONOTONIC, &t1);
+			printf("[info] indexed search %zd took: %ld\n", i, time_diff(t0, t1));
+			print_results(results);
+		}
+	} else if (type == "regex") {
+#if defined(LP_REGEX_HYPERSCAN)
+			scratch_cache.assign(num_threads, nullptr);
+#endif
+		for (size_t i = 0; i < searches.size(); i++) {
+			std::string& s = searches[i];
+			printf("[info] regex search %zd: %s\n", i, s.c_str());
+			std::vector<UniquePtr<LogParser::SearchResult>> results;
+#if defined(LP_REGEX_HYPERSCAN)
+			auto sample_db = RegexHS(s);
+			auto current = scratch_cache[0];
+			if (hs_alloc_scratch(sample_db.get_db(), &scratch_cache[0]) != HS_SUCCESS) {
+				perror("Cannot (re)allocate memory to search");
+				break;
+			}
+			// If a new allocation was done, clone the scratch to other threads
+			if (current != scratch_cache[0]) {
+				for (size_t i = 1; i < scratch_cache.size(); ++i) {
+					hs_free_scratch(scratch_cache[i]);
+					hs_clone_scratch(scratch_cache[0], &scratch_cache[i]);
+				}
+			}
```

Comment:
I'm confused about the purpose of this scratch cache. Could you link the docs for this? Hyperscan seems a have a weird API.

Reply:
This is actually a workaround to make Hyperscan work with current code. In [Hyperscan docs](http://intel.github.io/hyperscan/dev-reference/performance.html), scratch needs to be allocated for each thread ahead of time due to being expensive. I cannot put it anywhere else as the `search_regex` function always instantiates new Regex object. And I think it is better to cache it since it can be potentially reused without reallocation.